build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/BwcVersions.java

```diff
@@ -0,0 +1,319 @@
+        assertCurrentVersionMatchesParsed(currentVersionProperty);
+
+        this.unreleased = computeUnreleased();
+    }
+
+    // Visible for testing
+    BwcVersions(List<String> versionLines, Version currentVersionProperty) {
+        this(currentVersionProperty, parseVersionLines(versionLines));
+    }
+
+    private static List<VersionPair> parseVersionLines(List<String> versionLines) {
+        return versionLines.stream()
+            .map(LINE_PATTERN::matcher)
+            .filter(Matcher::matches)
+            .map(
+                match -> new VersionPair(
+                    new Version(Integer.parseInt(match.group(1)), Integer.parseInt(match.group(2)), Integer.parseInt(match.group(3))),
+                    new Version(Integer.parseInt(match.group(5)), Integer.parseInt(match.group(6)), Integer.parseInt(match.group(7)))
+                )
+            )
+            .sorted()
+            .toList();
+    }
+
+    private void assertCurrentVersionMatchesParsed(Version currentVersionProperty) {
+        if (currentVersionProperty.equals(currentVersion.elasticsearch) == false) {
+            throw new IllegalStateException(
+                "Parsed versions latest version does not match the one configured in build properties. "
+                    + "Parsed latest version is "
+                    + currentVersion.elasticsearch
+                    + " but the build has "
+                    + currentVersionProperty
+            );
+        }
+    }
+
+    /**
+     * Returns info about the unreleased version, or {@code null} if the version is released.
+     */
+    public UnreleasedVersionInfo unreleasedInfo(Version version) {
+        return unreleased.get(version);
+    }
+
+    public void forPreviousUnreleased(Consumer<UnreleasedVersionInfo> consumer) {
+        filterSupportedVersions(
+            getUnreleased().stream().filter(version -> version.equals(currentVersion.elasticsearch) == false).collect(Collectors.toList())
+        ).stream().map(unreleased::get).forEach(consumer);
+    }
+
+    private String getBranchFor(Version version) {
+        if (version.equals(currentVersion.elasticsearch)) {
+            // Just assume the current branch is 'master'. It's actually not important, we never check out the current branch.
+            return "master";
+        } else {
+            return version.getMajor() + "." + version.getMinor();
+        }
+    }
+
+    private Map<Version, UnreleasedVersionInfo> computeUnreleased() {
+        Set<VersionPair> unreleased = new TreeSet<>();
+        // The current version is being worked, is always unreleased
+        unreleased.add(currentVersion);
+        // Recurse for all unreleased versions starting from the current version
+        addUnreleased(unreleased, currentVersion, 0);
+
+        // Grab the latest version from the previous major if necessary as well, this is going to be a maintenance release
+        VersionPair maintenance = versions.stream()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It is more efficient to directly use Stream::min or Stream::max rather than Stream::sorted and Stream::findFirst.  The former has O(n) time complexity while the latter has higher time complexity.